File: dotnet/SDL.DXA.Modules.CampaignContent/Provider/CampaignAssetProvider.cs
using Sdl.Web.Common;
using Sdl.Web.Common.Configuration;
using Sdl.Web.Common.Logging;
using Sdl.Web.Common.Models;
using SDL.DXA.Modules.CampaignContent.Models;
using System;
using System.Collections.Concurrent;
using System.IO;
using System.IO.Compression;
using System.Threading;
using System.Web;
using System.Web.Configuration;
using Tridion.ContentDelivery.Meta;

namespace SDL.DXA.Modules.CampaignContent.Provider
{
    /// <summary>
    /// Campaign Asset Provider
    /// </summary>
    public class CampaignAssetProvider
    {
        private const string StaticsFolder = "BinaryData";

        private static CampaignAssetProvider _instance = null;

        private static readonly ConcurrentDictionary<string, CampaignContentMarkup> CachedMarkup = new ConcurrentDictionary<string, CampaignContentMarkup>();

        private static readonly ConcurrentDictionary<string, SemaphoreSlim> Semaphores = new ConcurrentDictionary<string, SemaphoreSlim>();

        // Cache time to keep the ZIP file in staging sites.
        // This to avoid to have the ZIP file unzipped for each request on a XPM enabled staging site.
        // 
        private readonly int stagingCacheTime = 0;

        private CampaignAssetProvider()
        {
            var cacheTime = WebConfigurationManager.AppSettings["instant-campaign-staging-cache-time"];
            if (cacheTime != null)
            {
                stagingCacheTime = Int32.Parse(cacheTime);
            }
        }

        public static CampaignAssetProvider Instance
        {
            get
            {
                if ( _instance == null)
                {
                    _instance = new CampaignAssetProvider();
                }
                return _instance;
            }
        }

        /// <summary>
        /// Get physical file name for a specific campaign asset
        /// </summary>
        /// <param name="localization"></param>
        /// <param name="campaignId"></param>
        /// <param name="assetUrl"></param>
        /// <returns></returns>
        public string GetAssetFileName(Localization localization, string campaignId, string assetUrl)
        {
            var assetFileName = GetBaseDir(localization, campaignId) + "/" + assetUrl;
            return assetFileName;
        }

        /// <summary>
        /// Get markup for a specific campaign
        /// </summary>
        /// <param name="campaignContentZip"></param>
        /// <param name="localization"></param>
        /// <returns></returns>
        public CampaignContentMarkup GetCampaignContentMarkup(CampaignContentZIP campaignContentZip, Localization localization)
        {
            return LoadCampaignContentMarkup(campaignContentZip, campaignContentZip.Id, localization);
        }

        /// <summary>
        /// Get markup of a specific campaign ID.
        /// </summary>
        /// <param name="campaignId"></param>
        /// <param name="localization"></param>
        /// <returns></returns>
        public CampaignContentMarkup GetCampaignContentMarkup(string campaignId, Localization localization)
        {
            return LoadCampaignContentMarkup(null, campaignId, localization);
        }

        /// <summary>
        /// Load markup for a specific campaign
        /// </summary>
        /// <param name="campaignContentZip"></param>
        /// <param name="campaignId"></param>
        /// <param name="localization"></param>
        /// <returns></returns>
        protected CampaignContentMarkup LoadCampaignContentMarkup(CampaignContentZIP campaignContentZip, string campaignId, Localization localization)
        {
            CampaignContentMarkup campaignContentMarkup;
            string cacheKey = GetMarkupCacheKey(campaignId, localization);
            CachedMarkup.TryGetValue(cacheKey, out campaignContentMarkup);
            string campaignBaseDir = GetBaseDir(localization, campaignId);

            if (campaignContentMarkup != null)
            {
                var zipFileLastModified = File.GetLastWriteTime(campaignContentMarkup.ZipFileName);
                if (!localization.IsXpmEnabled && zipFileLastModified > campaignContentMarkup.LastModified ||
                    localization.IsXpmEnabled && campaignContentMarkup.LastModified.AddSeconds(stagingCacheTime) < DateTime.Now)
                {
                    Log.Info("Zip has changed. Extracting campaign " + campaignId + ", last modified = " + zipFileLastModified);

                    var fileLock = GetFileSemaphore(campaignBaseDir);

                    // If no other thread is already working on it. Else keep using the existing version until the new one is available.
                    //
                    if (fileLock.CurrentCount > 0)
                    {
                        using (fileLock.UseWait())
                        {
                            try
                            {
                                var zipItem = campaignContentZip != null ? GetZipItem(campaignContentZip, localization) : GetZipItem(campaignId, localization);
                                if (ExtractZip(zipItem.ContentItem, campaignBaseDir, zipItem.ContentItem.LastModified))
                                {
                                    campaignContentMarkup = GetMarkup(campaignBaseDir);
                                    campaignContentMarkup.LastModified = zipItem.ContentItem.LastModified;
                                    campaignContentMarkup.ZipFileName = zipItem.LocalFileName;
                                    CachedMarkup[cacheKey] = campaignContentMarkup;
                                }
                            }
                            catch (Exception e)
                            {
                                Log.Error($"Could not get/extract ZIP file. Retry next time the campaign is accessed. \nException: {e}");
                            }
                        }
                    }
                }
            }
            else
            {
                var fileLock = GetFileSemaphore(campaignBaseDir);
                using (fileLock.UseWait())
                {
                    var zipItem = campaignContentZip != null ? GetZipItem(campaignContentZip, localization) : GetZipItem(campaignId, localization);
                    if (!Directory.Exists(campaignBaseDir) || Directory.GetFiles(campaignBaseDir).Length == 0)
                    {
                        Log.Info("Extracting campaign " + campaignId + ", last modified = " + zipItem.ContentItem.LastModified);
                        if (!ExtractZip(zipItem.ContentItem, campaignBaseDir, zipItem.ContentItem.LastModified))
                        {
                            throw new DxaException($"Could not extract campaign with ID: {campaignId}");
                        }
                    }

                    campaignContentMarkup = GetMarkup(campaignBaseDir);
                    campaignContentMarkup.LastModified = zipItem.ContentItem.LastModified;
                    campaignContentMarkup.ZipFileName = zipItem.LocalFileName;
                    CachedMarkup[cacheKey] = campaignContentMarkup;
                }
            }

            return campaignContentMarkup;
        }

        /// <summary>
        /// Get last modified date for a specific campaign and localization.
        /// </summary>
        /// <param name="campaignId"></param>
        /// <param name="localization"></param>
        /// <returns></returns>
        public DateTime GetLastModified(string campaignId, Localization localization)
        {
            string cacheKey = GetMarkupCacheKey(campaignId, localization);
            CampaignContentMarkup campaignContentMarkup;
            CachedMarkup.TryGetValue(cacheKey, out campaignContentMarkup);
            if ( campaignContentMarkup != null)
            {
                return campaignContentMarkup.LastModified;
            }
            else
            {
                return DateTime.Now;
            }         
        }

        /// <summary>
        /// Get markup cache key
        /// </summary>
        /// <param name="campaignId"></param>
        /// <param name="localization"></param>
        /// <returns></returns>
        private string GetMarkupCacheKey(string campaignId, Localization localization)
        {
            return campaignId + "-" + localization.Id;
        }

        /// <summary>
        /// Get ZIP item.
        /// </summary>
        /// <param name="campaignContentZip"></param>
        /// <returns></returns>
        protected ZipItem GetZipItem(CampaignContentZIP campaignContentZip, Localization localization)
        {
            var item = SiteConfiguration.ContentProvider.GetStaticContentItem(campaignContentZip.Url, localization);
            return new ZipItem
            {
                ContentItem = item,
                UrlPath = campaignContentZip.Url,
                LocalFileName = GetBinaryLocalFileName(campaignContentZip.Url, localization)
            };
        }

        protected ZipItem GetZipItem(string itemId, Localization localization)
        {
            BinaryMetaFactory binaryMetaFactory = new BinaryMetaFactory();
            BinaryMeta binaryMeta = binaryMetaFactory.GetMeta("tcm:" + localization.Id + "-" + itemId);
            if (binaryMeta != null)
            {
                var item = SiteConfiguration.ContentProvider.GetStaticContentItem(binaryMeta.UrlPath, localization);
                return new ZipItem
                {
                    ContentItem = item,
                    UrlPath = binaryMeta.UrlPath,
                    LocalFileName = GetBinaryLocalFileName(binaryMeta.UrlPath, localization)
                }; 
            }
            return null;
        }

        /// <summary>
        /// Get binary local file of a binary with a given URL path and localization.
        /// </summary>
        /// <param name="urlPath"></param>
        /// <param name="localization"></param>
        /// <returns></returns>
        protected string GetBinaryLocalFileName(string urlPath, Localization localization)
        {
            string baseDir = AppDomain.CurrentDomain.BaseDirectory;
            string localFilePath = $"{baseDir}/{localization.BinaryCacheFolder}/{urlPath}";
            return localFilePath;
        }

        /// <summary>
        /// Get base directory for the specified campaign ID
        /// </summary>
        /// <param name="localization"></param>
        /// <param name="campaignId"></param>
        /// <returns></returns>
        protected string GetBaseDir(Localization localization, string campaignId)
        {
            var webAppBaseDir = HttpContext.Current.Server.MapPath("~/");
            var baseDir = webAppBaseDir + GetLocalStaticsFolder(localization.Id) + "/campaign-content/" + campaignId;
            return baseDir;           
        }

        /// <summary>
        /// Extract campaign ZIP to specified directory.
        /// </summary>
        /// <param name="zipItem"></param>
        /// <param name="directory"></param>
        /// <param name="zipLastModified"></param>
        protected bool ExtractZip(StaticContentItem zipItem, String directory, DateTime zipLastModified)
        { 
            if (Directory.Exists(directory))
            {
                if (Directory.GetCreationTime(directory) > zipLastModified && Directory.GetFiles(directory).Length > 0)
                {
                    Log.Info("Campaign assets in directory '" + directory + "' is already up to date. Skipping recreation of campaign assets.");
                    return false;
                }

                try
                {
                    Directory.Delete(directory, true);
                }
                catch (Exception e)
                {
                    Log.Warn($"Could not delete cached campaign resources in: {directory}. \nException: {e}");
                }
            }

            if (!Directory.Exists(directory))
            {
                Directory.CreateDirectory(directory);
            }

            try
            {
                using (ZipArchive archive = new ZipArchive(zipItem.GetContentStream()))
                {
                    archive.ExtractToDirectory(directory);
                }
            }
            catch (Exception e)
            {
                Log.Warn($"Could not unzip campaign resources in: {directory}. Will rely on the current content there. \nException: {e}");
                return false;
            }
            return true;
        }

        /// <summary>
        /// Get campaign markup
        /// </summary>
        /// <param name="directory"></param>
        /// <returns></returns>
        protected CampaignContentMarkup GetMarkup(string directory)
        {
            var markup = new CampaignContentMarkup();

            // Main HTML
            //
            string mainHtmlFile = directory + "/index.html";
            if (File.Exists(mainHtmlFile))
            {
                markup.MainHtml = ReadAllTextWithRetry(mainHtmlFile);
            }

            // Header HTML
            //
            string headerHtmlFile = directory + "/header.html";
            if (File.Exists(headerHtmlFile))
            {
                markup.HeaderHtml = ReadAllTextWithRetry(headerHtmlFile);
            }

            // Footer HTML
            //
            string footerHtmlFile = directory + "/footer.html";
            if (File.Exists(footerHtmlFile))
            {
                markup.FooterHtml = ReadAllTextWithRetry(footerHtmlFile);
            }
            return markup;
        }


        protected static SemaphoreSlim GetFileSemaphore(string name)
        {
            return Semaphores.GetOrAdd(name, s => new SemaphoreSlim(1));
        }

        private static string GetLocalStaticsFolder(string localizationId)
        {
            return string.Format("{0}\\{1}", StaticsFolder, localizationId);
        }

        private static string ReadAllTextWithRetry(string path, int triesLeft = 3, TimeSpan? delay = null)
        {
            delay = delay ?? TimeSpan.FromMilliseconds(10);

            try
            {
                return File.ReadAllText(path);
            }
            catch (IOException)
            {
                if (triesLeft <= 0)
                {
                    throw;
                }

                Thread.Sleep(delay.Value);

                return ReadAllTextWithRetry(path, --triesLeft, delay);
            }
        }
    }


    // Extensions to SemaphoreSlim to make it possible to use it using blocks
    //
    static class SemaphoreSlimExtensions
    {
       public static IDisposable UseWait(
       this SemaphoreSlim semaphore)
        {
            semaphore.Wait();
            return new ReleaseWrapper(semaphore);
        }

        private class ReleaseWrapper : IDisposable
        {
            private readonly SemaphoreSlim _semaphore;

            private bool _isDisposed;

            public ReleaseWrapper(SemaphoreSlim semaphore)
            {
                _semaphore = semaphore;
            }

            public void Dispose()
            {
                if (_isDisposed)
                    return;

                _semaphore.Release();
                _isDisposed = true;
            }
        }
    }
}

File: java/campaigncontent-dxa-module/src/main/java/com/sdl/dxa/modules/campaigncontent/provider/CampaignAssetProvider.java
package com.sdl.dxa.modules.campaigncontent.provider;

import com.sdl.dxa.modules.campaigncontent.model.CampaignContentZIP;
import com.sdl.webapp.common.api.content.ContentProvider;
import com.sdl.webapp.common.api.content.ContentProviderException;
import com.sdl.webapp.common.api.content.StaticContentItem;
import com.sdl.webapp.common.api.localization.Localization;
import com.tridion.meta.BinaryMeta;
import com.tridion.meta.BinaryMetaFactory;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.context.WebApplicationContext;

import java.io.*;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

/**
 * Campaign Asset Provider
 */
@Component
@Slf4j
public class CampaignAssetProvider {

    @Autowired
    private ContentProvider contentProvider;

    @Autowired
    private WebApplicationContext webApplicationContext;

    private Map<String, CampaignContentMarkup> cachedMarkup = new HashMap<>();

    // Cache time to keep the ZIP file in staging sites.
    // This to avoid to have the ZIP file unzipped for each request on a XPM enabled staging site.
    //
    @Value("${instantcampaign.staging.cacheTime:0}")
    private int stagingCacheTime;

    static final String STATIC_FILES_DIR = "BinaryData";

    /**
     * Get a specific asset in a campaign
     * @param localization
     * @param campaignId
     * @param assetUrl
     * @return asset input stream. If not found -> NULL is returned.
     * @throws ContentProviderException
     */
    public InputStream getAsset(Localization localization, String campaignId, String assetUrl) throws ContentProviderException {

        try {
            log.debug("Getting campaign asset: " + assetUrl + " for campaign ID: " + campaignId);
            File assetFile = new File(getBaseDir(localization, campaignId) + assetUrl);
            if (!assetFile.exists()) {

                // If asset is not available -> trigger unpack the ZIP file & make the campaign markup available.
                // This is to support LB scenario when one node unpacks the ZIP file while other
                // nodes receives the asset requests.
                //
                StaticContentItem zipItem = getZipItem(campaignId, localization);
                if (zipItem == null) {

                    // Campaign asset was not found
                    //
                    return null;
                }
                else {
                    // Trigger unpack of the ZIP file
                    //
                    getCampaignContentMarkup(campaignId, zipItem, localization);
                }
                if (!assetFile.exists()) {
                    log.warn("Campaign asset '" + assetUrl + "' was not found for campaign ID: " + campaignId);
                    return null;
                }
            }
            return new FileInputStream(new File(getBaseDir(localization, campaignId) + assetUrl));
        }
        catch ( IOException e ) {
            throw new ContentProviderException("Could not get asset: " + assetUrl + " for campaign: " + campaignId + " and localication ID: " + localization.getId(), e);
        }

    }

    /**
     * Get the markup (main, header, footer etc) for campaign.
     * @param campaignContentZip
     * @param localization
     * @return markup
     * @throws ContentProviderException
     */
    public CampaignContentMarkup getCampaignContentMarkup(CampaignContentZIP campaignContentZip, Localization localization) throws ContentProviderException {
        StaticContentItem zipItem = this.getZipItem(campaignContentZip, localization);
        return this.getCampaignContentMarkup(campaignContentZip.getId(), zipItem, localization);
    }

    public CampaignContentMarkup getCampaignContentMarkup(String campaignId, StaticContentItem zipItem, Localization localization) throws ContentProviderException {
        String cacheKey = getMarkupCacheKey(campaignId, localization);
        CampaignContentMarkup markup = this.cachedMarkup.get(cacheKey);
        File baseDir = this.getBaseDir(localization, campaignId);

        if( markup == null ||
                !baseDir.exists() ||
                !this.directoryHasFiles(baseDir) ||
                !localization.isStaging() && zipItem.getLastModified() > markup.getLastModified() ||
                localization.isStaging() && markup.getLastModified()+(stagingCacheTime*1000) < System.currentTimeMillis()) {
            try {
                markup = extractZip(zipItem, baseDir);
                markup.setLastModified(zipItem.getLastModified());
                this.cachedMarkup.put(cacheKey, markup);
            }
            catch ( IOException e ) {
                throw new ContentProviderException("Could not extract free format ZIP file.", e);
            }
        }
        return markup;
    }

    /**
     * Get last modified time on the campaign assets (not the content)
     * @param campaignId
     * @param localization
     * @return
     */
    public long getLastModified(String campaignId, Localization localization) {
        CampaignContentMarkup markup = this.cachedMarkup.get(getMarkupCacheKey(campaignId, localization));
        if ( markup != null ) {
            return markup.getLastModified();
        }
        else {
            return -1;
        }
    }

    private String getMarkupCacheKey(String campaignId, Localization localization) {
        return campaignId + "-" + localization.getId();
    }

    /**
     * Get physical base directory for the campaign content
     * @param localization
     * @param campaignId
     * @return base directory
     */
    protected File getBaseDir(Localization localization, String campaignId) {
        return new File(StringUtils.join(new String[]{
                webApplicationContext.getServletContext().getRealPath("/"), STATIC_FILES_DIR, localization.getId(), "campaign-content", campaignId
        }, File.separator));
    }

    /**
     * Get ZIP item
     * @param campaignContentZip
     * @param localization
     * @return
     * @throws ContentProviderException
     */
    protected StaticContentItem getZipItem(CampaignContentZIP campaignContentZip, Localization localization) throws ContentProviderException {
        return contentProvider.getStaticContent(campaignContentZip.getUrl(), localization.getId(), localization.getPath());
    }

    protected StaticContentItem getZipItem(String itemId, Localization localization) throws ContentProviderException {
        BinaryMetaFactory binaryMetaFactory = new BinaryMetaFactory();
        BinaryMeta binaryMeta = binaryMetaFactory.getMeta("tcm:" + localization.getId() + "-" + itemId);
        if (binaryMeta != null) {
            return contentProvider.getStaticContent(binaryMeta.getURLPath(), localization.getId(), localization.getPath());
        }
        return null;
    }

    /**
     * Extract ZIP
     * @param zipItem
     * @param directory
     * @return HTML fragments
     * @throws IOException
     */
    protected CampaignContentMarkup extractZip(StaticContentItem zipItem, File directory) throws IOException {

        log.debug("Extracting campaign ZIP in directory: " + directory);
        ByteArrayOutputStream htmlMarkup = new ByteArrayOutputStream();
        ByteArrayOutputStream headerMarkup = new ByteArrayOutputStream();
        ByteArrayOutputStream footerMarkup = new ByteArrayOutputStream();

        byte[] buffer = new byte[1024];

        directory.mkdir();

        // Get the zip file content
        //
        ZipInputStream zis =
                new ZipInputStream(zipItem.getContent());

        // Get the zipped file list entry
        //
        ZipEntry ze = zis.getNextEntry();

        while (ze!=null) {
            String fileName = ze.getName();
            File newFile = new File(directory + File.separator + fileName);

            // Create all non exists folders
            // else you will hit FileNotFoundException for compressed folder
            //
            if (ze.isDirectory()) {
                newFile.mkdirs();
            }
            else {
                newFile.getParentFile().mkdirs();
                newFile.createNewFile();
                FileOutputStream fos = new FileOutputStream(newFile);
                int len;
                while ((len = zis.read(buffer)) > 0) {
                    fos.write(buffer, 0, len);
                    if (fileName.equals("index.html") ) {
                        htmlMarkup.write(buffer, 0, len);
                    }
                    else if (fileName.equals("header.html") ) {
                        headerMarkup.write(buffer, 0, len);
                    }
                    else if (fileName.equals("footer.html") ) {
                        footerMarkup.write(buffer, 0, len);
                    }

                }
                fos.close();
            }
            ze = zis.getNextEntry();
        }
        zis.closeEntry();
        zis.close();

        CampaignContentMarkup campaignContentMarkup = new CampaignContentMarkup();
        if ( htmlMarkup.size() > 0 ) {
            campaignContentMarkup.setMainHtml(new String(htmlMarkup.toByteArray()));
        }
        if ( headerMarkup.size() > 0 ) {
            campaignContentMarkup.setHeaderHtml(new String(headerMarkup.toByteArray()));
        }
        if ( footerMarkup.size() > 0 ) {
            campaignContentMarkup.setFooterHtml(new String(footerMarkup.toByteArray()));
        }
        return campaignContentMarkup;
    }

    /**
     * Check if a specific directory has files
     * @param directory
     * @return
     */
    protected boolean directoryHasFiles(File directory) {
        String[] files = directory.list();
        return files != null && files.length > 0;

    }

}
